src/lib.rs

```diff
@@ -40,19 +42,358 @@ mod repl_helper;
+        ///
+        /// # Parameters
+        /// * `machine`: a mutable reference to a `Machine` instance.
+        /// * `query_state`: a raw pointer to a `QueryState` instance.
+        ///
+        /// # Returns
+        /// - Returns a raw pointer to a C-string containing the serialized JSON result.
+        ///
+        /// # Expected Response Format
+        /// ```json
+        /// // if result is a binding
+        /// // current limitation is that only concrete (equality) bindings are returned,
+        /// // residual goals not yet supported.
+        /// {
+        ///   "status": "ok",  // Can also be "error" or "panic"
+        ///   "result": [{ ... }], // single map entry, otherwise same signature as `scryer_run_query()`
+        /// }
+        /// 
+        /// // if result is a boolean goal
+        /// 
+        /// {
+        ///   "status": "ok",  // Can also be "error" or "panic"
+        ///   "result": boolean // note -- `scryer_run_query_generator()` will continue returning results if
+        ///                     //       .. successively invoked, but the result will always be the same
+        ///                     //       .. at this point, `scryer_cleanup_query_generator()` should be called 
+        /// }        
+        /// 
+        /// // if panic
+        /// { 
+        ///   "status": "error" | "panic",
+        ///   "error": error message | "panic"
+        /// }
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_run_query_generator(machine: &mut Machine, query_state: *mut QueryState) -> *mut c_char {
+            let query_state = unsafe { &mut *(query_state) };
+            let query_resolution = machine.run_query_generator(query_state);
+            let value: serde_json::Value = serde_json::from_str(&format!("{}", query_resolution.expect("Error while marshaling JSON"))).unwrap();
+            let output_string = serde_json::to_string(&serde_json::json!({"status": "ok", "result": value})).unwrap();
+            let c_string = CString::new(output_string).unwrap();
+            c_string.into_raw()
+        }
+
+        /// Loads a Scryer Prolog module from a string.
+        ///
+        /// # Safety
+        ///
+        /// - `machine` must be a valid mutable pointer to a `Machine`.
+        /// - `input` must be a valid pointer to a null-terminated C string.
+        ///
+        /// # Arguments
+        ///
+        /// * `machine` - A mutable reference to the `Machine` to load the module into.
+        /// * `input` - A pointer to a null-terminated C string containing the module source code.
+        ///
+        /// # Returns
+        ///
+        /// - `0` if the module was successfully loaded.
+        /// - `1` if an error occurred while loading the module.
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_load_module_string(machine: &mut Machine, input: *const c_char) -> i32 {
```

Comment:
A benefit of having `i32` there instead is that if we need more kinds of errors later it would still be ABI compatible. But yeah, I don't think there will be more possibilities for this one.